sdk/samples/iot/paho_iot_hub_pnp_sample.c

```diff
@@ -66,954 +49,917 @@ static az_span boot_time_span;
-              &client,
-              request_id_span,
-              reported_property_topic,
-              sizeof(reported_property_topic),
-              NULL)))
+  // Parse the incoming message topic and check which feature it is for.
+  if (az_succeeded(
+          rc
+          = az_iot_hub_client_twin_parse_received_topic(&hub_client, topic_span, &twin_response)))
   {
-    printf("Unable to get twin document publish topic, return code %d\n", rc);
-    return rc;
-  }
+    LOG_SUCCESS("Client received a valid topic response:");
+    LOG_AZ_SPAN("Topic:", topic_span);
+    LOG_AZ_SPAN("Payload:", message_span);
+    LOG("Status: %d", twin_response.status);
 
-  // Twin reported properties must be in JSON format. The payload is constructed here.
-  az_json_writer json_builder;
-  if (is_max_reported_prop)
+    handle_device_twin_message(message_span, &twin_response);
+  }
+  else if (az_succeeded(az_iot_hub_client_methods_parse_received_topic(
+               &hub_client, topic_span, &command_request)))
   {
-    if (az_failed(
-            rc
-            = build_reported_property(&json_builder, max_temp_reported_property_name, temp_value)))
-    {
-      return rc;
-    }
+    LOG_SUCCESS("Client received a valid topic response:");
+    LOG_AZ_SPAN("Topic:", topic_span);
+    LOG_AZ_SPAN("Payload:", message_span);
+
+    handle_command_message(message_span, &command_request);
   }
   else
   {
-    if (az_failed(
-            rc = build_confirmed_reported_property(
-                &json_builder,
-                desired_temp_property_name,
-                temp_value,
-                200,
-                version,
-                AZ_SPAN_FROM_STR("success"))))
-    {
-      return rc;
-    }
+    LOG_ERROR("Message from unknown topic: az_result return code 0x%04x.", rc);
+    LOG_AZ_SPAN("Topic:", topic_span);
+    exit(rc);
+  }
+}
+
+static void handle_device_twin_message(
+    const az_span twin_message_span,
+    const az_iot_hub_client_twin_response* twin_response)
+{
+  // Invoke appropriate action per response type (3 Types only).
+  bool is_twin_get = false;
+
+  switch (twin_response->response_type)
+  {
+    case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_GET:
+      LOG("Type: GET");
+      is_twin_get = true;
+      break;
+
+    case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES:
+      LOG("Type: Desired Properties");
+      break;
+
+    case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_REPORTED_PROPERTIES:
+      LOG("Type: Reported Properties");
+      break;
   }
-  az_span json_payload = az_json_writer_get_bytes_used_in_destination(&json_builder);
 
-  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+  // For a GET response OR a Desired Properties response from the server:
+  // 1. Parse for the desired temperature
+  // 2. Update device temperature locally
+  // 3. Report updated temperature to server
+  double desired_temp = 0.0;
+  int32_t version_num = 0;
```

Comment:
Using a variable without initialization is a bad thing though.  I've always known the best practice when it comes to variables is to initialize immediately.  Confirmed this with Yuxin a while back.

Reply:
What if these are set to something unreasonable, so if they were to be used for some reason, we would know they weren't set, rather than seeing garbage.
version_num = -1